instrumentation/aws_sdk/lib/opentelemetry/instrumentation/aws_sdk/handler.rb

```diff
@@ -9,28 +9,37 @@ module Instrumentation
     module AwsSdk
       # Generates Spans for all interactions with AwsSdk
       class Handler < Seahorse::Client::Handler
-        def call(context) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+        SQS_SEND_MESSAGE = 'SQS.SendMessage'
+        SQS_SEND_MESSAGE_BATCH = 'SQS.SendMessageBatch'
+        SQS_RECEIVE_MESSAGE = 'SQS.ReceiveMessage'
+        SNS_PUBLISH = 'SNS.Publish'
+
+        def call(context) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
           return super unless context
 
-          service_name = context.client.class.api.metadata['serviceId'] || context.client.class.to_s.split('::')[1]
+          service_name = context.client.class.to_s.split('::')[1]
           operation = context.operation&.name
+          client_method = "#{service_name}.#{operation}"
           attributes = {
             'aws.region' => context.config.region,
             OpenTelemetry::SemanticConventions::Trace::RPC_SYSTEM => 'aws-api',
             OpenTelemetry::SemanticConventions::Trace::RPC_METHOD => operation,
             OpenTelemetry::SemanticConventions::Trace::RPC_SERVICE => service_name
           }
           attributes[SemanticConventions::Trace::DB_SYSTEM] = 'dynamodb' if service_name == 'DynamoDB'
+          MessagingHelper.apply_sqs_attributes(attributes, context, client_method) if service_name == 'SQS'
+          MessagingHelper.apply_sns_attributes(attributes, context, client_method) if service_name == 'SNS'
 
-          tracer.in_span("#{service_name}.#{operation}", attributes: attributes, kind: OpenTelemetry::Trace::SpanKind::CLIENT) do |span|
+          tracer.in_span(span_name(context, client_method), attributes: attributes, kind: span_kind(operation)) do |span|
+            inject_context(context, client_method)
```

Comment:
Please provide a way to opt-out of automatic context propagation. We have internal systems which are doing context propagation within the message body and we'd rather not duplicate that in attributes.

Reply:
@YanivD take a look at how other instrumentations do this, e.g. ActiveJob provides link, child, or none

https://github.com/open-telemetry/opentelemetry-ruby/blob/main/instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/instrumentation.rb#L55

Reply:
@johnduhart 
Thanks. I think it's a good idea.
I'm thinking how to call the option. Suggestions are very welcome.

@arielvalentin 
In this PR I didn't add support for `process` spans which extract the context propagation. I'm not sure `propagation_style` is the best fit. WDYT?

Reply:
My mistake! I did not look closely enough to see the PR only injects context. 

Is there a separate PR for extraction?

Reply:
Agreed, we absolutely need to supply

- an `option` for enabling(as child or as link) / disabling propagation as @johnduhart points out. 
- support for both injecting and extracting context as @arielvalentin points out.

I think one of the difficulties here is that `SQS_RECEIVE_MESSAGE` can receive either a single message, or a batch of messages. 

- The simple case, the default, which is we receive, at maximum, `1` message: We can infer this by looking at the [`MaxNumberOfMessages` request parameter as defined here](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/APIReference/API_ReceiveMessage.html), which is how our [good friends at Datadog have approached this](https://github.com/DataDog/dd-trace-js/blob/25264dce67a97384d7f1abc541155cecc2324679/packages/datadog-plugin-aws-sdk/src/services/sqs.js#L45-L48). In the case where we are receiving 1 message, we can relatively easily choose to extract as a child, as a link, or not at all. We can follow the approach taken, as @arielvalentin points out, in [one of our other instruments, like ActiveJob](https://github.com/open-telemetry/opentelemetry-ruby/blob/871ddb4086624db134e003a55e9c57e687bffe25/instrumentation/active_job/lib/opentelemetry/instrumentation/active_job/patches/active_job_callbacks.rb#L30-L55).
- The complex case, which isn't default but practically speaking is the more common pattern since it's more efficient, which is that we can receive _many_ messages in a `RECEIVE_MESSAGE` `handler`, each message potentially containing a different trace-id/parent-span. In this case, because we don't have a seperate span for processing _each message_, we can't effectively extract context in a parent/child relationship since each individual message has it's own parent. but our span can only have 1 parent.  So we're at a bit of an impasse since we need to handle this in some way.  Practically speaking, I think there are three options here:
  - First,  _the datadog way_, where we could simply just _not support_ propagation of trace context when we detect a potential a "batch" of messages ( `MaxNumberOfMessages > 1` ). We aren't really beholden to any specification here since this instrumentation is experimental, so i think if push comes to shove I don't hate this approach, but I think it's practically speaking a _"Half Measure"_, as Mike Ehrmantraut might say.
  - Second, _the node/aspecto way_ , where we create a _span per message processed_, and then because we're back to a 1:1 relationship with span => parent, we can easily choose to propagate as `child/link/not-at-all`. I think this would require additional monkey-patching of the underlying aws-sdk ruby library, so that we can measure each individual message processed, and i haven't reviewed this library closely enough to know if/how-complex that is to do, but that seems like that would match the functionality of Node, and also allow us to expose the option like @arielvalentin is describing and like i've described in the "simple" case.
  - Third, _the Links way_, where we simply do not do any additional monkey patching, but, when we detect that `MaxNumberOfMessages > 1`, we iterate over each message's headers, extract each valid trace context, and attach them _all_ as Span Links of the span.  I actually tend to think this is the "semantically correct" use of [opentelemetry span links](https://github.com/open-telemetry/opentelemetry-specification/blob/e2c2472985b17e37a25a7dc5aa0aa071e6683c98/specification/overview.md#links-between-spans), but since practically speaking many backends don't support Span Links, would still result in disconnected traces in practice, and also wouldn't allow people to maintain parent/child relationships for their spans. This would be a change in convention from both how Datadog and how Node/Aspecto do things currently, but is perhaps a middle ground between the half measure of the 1st option i suggested, and the extra work of the 2nd option i suggested.


So, I think we need to make a determination here around how we want to handle this. I'm open to feedback here, but IMO both of these details (options to enable/disable, and how to support extract propagation in the simple/complex case) I think need to be resolved before we can support sqs propagation and merge this PR.

Reply:
@arielvalentin
I'm not planning to add support for message context extraction soon. 
There are few open issues with messaging spec for processing SQS messages. [Link for discussion](https://github.com/open-telemetry/opentelemetry-ruby/discussions/1057) about some of the issues. I'll probably implement extraction in the next few months after some messaging spec changes.